File: sdk/platform/Web/src/ArucoDetection.cpp
#include <Posemesh/ArucoDetection.hpp>
#include <emscripten/bind.h>
#include <memory>

using namespace emscripten;
using namespace psm;

namespace {
bool detectArucoFromLuminance(
    const std::vector<uint8_t>& imageBytes,
    int width,
    int height,
    int markerFormat,
    std::vector<std::string>& outContents,
    std::vector<std::shared_ptr<Vector2>>& outCorners)
{
    std::vector<Vector2> corners;
    bool detectionResult = ArucoDetection::detectArucoFromLuminance(imageBytes.data(), imageBytes.size(), width, height, (ArucoMarkerFormat)markerFormat, outContents, corners);

    outCorners.clear();
    for (auto& corner : corners) {
        outCorners.emplace_back(new Vector2(std::move(corner)));
    }

    return detectionResult;
}

std::vector<std::shared_ptr<psm::LandmarkObservation>> detectArucoFromLuminanceLandmarkObservations(
    const std::vector<uint8_t>& imageBytes,
    int width,
    int height,
    int markerFormat)
{
    std::vector<std::shared_ptr<psm::LandmarkObservation>> observations;
    auto lo = ArucoDetection::detectArucoFromLuminance(imageBytes, width, height, (ArucoMarkerFormat)markerFormat);
    for (auto& observation : lo) {
        observations.emplace_back(new psm::LandmarkObservation(std::move(observation)));
    }

    return observations;
}
}

EMSCRIPTEN_BINDINGS(ArucoDetection)
{
    class_<ArucoDetection>("ArucoDetection")
        .class_function("__detectArucoFromLuminance(imageBytes, width, height, markerFormat, outContents, outCorners)", &detectArucoFromLuminance)
        .class_function("__detectArucoFromLuminanceLandmarkObservations(imageBytes, width, height, markerFormat)", &detectArucoFromLuminanceLandmarkObservations);
}

File: sdk/platform/Web/ArucoDetection.js
posemeshModule.ArucoDetection = null;

__internalPosemeshAPI.builderFunctions.push(function() {
    __internalPosemesh.ArucoDetection.detectArucoFromLuminance = function(imageBytes, width, height, markerFormat, outContents, outCorners) {
        let imageBytesVector = undefined, outContentsVector = undefined, outCornersVector = undefined;
        try {
            imageBytesVector = __internalPosemeshAPI.toVectorUint8(imageBytes);
            outContentsVector = new __internalPosemesh.VectorString();
            outCornersVector = new __internalPosemesh.VectorVector2();
            let result = __internalPosemesh.ArucoDetection.__detectArucoFromLuminance(imageBytesVector, width, height, markerFormat, outContentsVector, outCornersVector);

            if (result) {
                outContents.length = 0;
                let outContentsVectorSize = outContentsVector.size();
                for (let i = 0; i < outContentsVectorSize; i++) {
                    outContents.push(outContentsVector.get(i));
                }

                for (let corner of outCorners) {
                    corner.delete();
                }
                outCorners.length = 0;
                let outCornersVectorSize = outCornersVector.size();
                for (let i = 0; i < outCornersVectorSize; i++) {
                    outCorners.push(outCornersVector.get(i));
                }
            }

            return result;
        } finally {
            if (outCornersVector) {
                outCornersVector.delete();
            }
            if (outContentsVector) {
                outContentsVector.delete();
            }
            if (imageBytesVector) {
                imageBytesVector.delete();
            }
        }
    };

    __internalPosemesh.ArucoDetection.detectArucoFromLuminanceLandmarkObservations = function(imageBytes, width, height, markerFormat) {
        let imageBytesVector = undefined;
        try {
            imageBytesVector = __internalPosemeshAPI.toVectorUint8(imageBytes)
            let result = []
            let observations = __internalPosemesh.ArucoDetection.__detectArucoFromLuminanceLandmarkObservations(imageBytesVector, width, height, markerFormat)

            for (let i = 0; i < observations.size(); i++) {
                result.push(observations.get(i))
            }

            return result;
        } finally {
            if (imageBytesVector) {
                imageBytesVector.delete();
            }
        }
    };

    posemeshModule.ArucoDetection = __internalPosemesh.ArucoDetection;
});
